emr-datahub/src/main/scala/org/apache/spark/streaming/aliyun/datahub/DatahubIterator.scala

```diff
@@ -0,0 +1,112 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.streaming.aliyun.datahub
+
+import java.util.concurrent.LinkedBlockingQueue
+
+import scala.collection.JavaConversions._
+import com.aliyun.datahub.model.{OffsetContext, RecordEntry}
+import org.I0Itec.zkclient.ZkClient
+import org.apache.spark.TaskContext
+import org.apache.spark.internal.Logging
+import org.apache.spark.util.NextIterator
+
+private class DatahubIterator(
+    datahubClient: DatahubClientAgent,
+    endpoint: String,
+    project: String,
+    topic: String,
+    shardId: String,
+    cursor: String,
+    count: Int,
+    subId: String,
+    accessId: String,
+    accessKey: String,
+    func: RecordEntry => String,
+    zkClient: ZkClient,
+    checkpointDir: String,
+    context: TaskContext) extends NextIterator[Array[Byte]] with Logging{
+  private val datahubShardMaxSize = 512
+  private val step = 100
+  private var dataBuffer = new LinkedBlockingQueue[Array[Byte]](datahubShardMaxSize * step * 10)
+  private var hasRead = 0
```

Comment:
1. add a new config about 'step' and 'factor'
2. please review the logic about databuffer size again

Reply:
Initialize a queue with size "datahubShardMaxSize * step * 10"? IMHO, this should be "step" size. Right?